src/pkg/zoci/push.go

```diff
@@ -28,89 +29,119 @@ import (
-		err = errors.Join(err, err2)
-	}()
-
 	copyOpts := r.OrasRemote.GetDefaultCopyOpts()
-	copyOpts.Concurrency = concurrency
-	totalSize := oci.SumDescsSize(descs) + root.Size + manifestConfigDesc.Size
-	logger.From(ctx).Info("pushing package to registry", "destination", r.Repo().Reference.String(),
-		"architecture", pkgLayout.Pkg.Build.Architecture, "size", utils.ByteFormat(float64(totalSize), 2))
-
-	trackedRemote := images.NewTrackedTarget(r.Repo(), totalSize, images.DefaultReport(r.Log(), "package publish in progress", r.Repo().Reference.String()))
-	trackedRemote.StartReporting(ctx)
-	defer trackedRemote.StopReporting()
-	publishedDesc, err := oras.Copy(ctx, src, root.Digest.String(), trackedRemote, "", copyOpts)
+	copyOpts.Concurrency = opts.OCIConcurrency
+
+	// For progress reporting and size estimation
+	// (root + manifestConfigDesc sizes are unknown until built each attempt;
+	// this is a conservative total using layer sizes; progress still works fine.)
+	totalSize := oci.SumDescsSize(descs)
+
+	var publishedDesc ocispec.Descriptor
+
+	err = retry.Do(
+		func() error {
+			l.Info("pushing package to registry", "destination", r.Repo().Reference.String(),
+				"architecture", pkgLayout.Pkg.Build.Architecture, "size", utils.ByteFormat(float64(totalSize), 2))
+			manifestConfigDesc, cfgErr := r.OrasRemote.CreateAndPushManifestConfig(ctx, annotations, ZarfConfigMediaType)
+			if cfgErr != nil {
+				return cfgErr
+			}
+
+			root, packErr := r.OrasRemote.PackAndTagManifest(ctx, src, descs, manifestConfigDesc, annotations)
+			if packErr != nil {
+				return packErr
+			}
+			// Always remove the temp manifest file created by PackAndTagManifest
+			defer func() {
+				err2 := os.Remove(pkgLayout.Pkg.Metadata.Name)
+				err = errors.Join(err, err2)
+			}()
+
+			// Update the total with manifest + config for better progress (optional)
+			attemptTotal := totalSize + root.Size + manifestConfigDesc.Size
+
+			trackedRemote := images.NewTrackedTarget(
+				r.Repo(),
+				attemptTotal,
+				images.DefaultReport(r.Log(), "package publish in progress", r.Repo().Reference.String()),
+			)
+			trackedRemote.StartReporting(ctx)
+			defer trackedRemote.StopReporting()
+
+			publishedDesc, copyErr := oras.Copy(ctx, src, root.Digest.String(), trackedRemote, "", copyOpts)
+			if copyErr != nil {
+				return copyErr
+			}
+
+			return r.OrasRemote.UpdateIndex(ctx, r.Repo().Reference.Reference, publishedDesc)
+		},
+		retry.Attempts(uint(opts.Retries)),
+		retry.Delay(defaultDelayTime),
+		retry.MaxDelay(defaultMaxDelayTime),
+		retry.DelayType(retry.BackOffDelay), // exponential backoff
+		retry.LastErrorOnly(true),
+		retry.Context(ctx),
+		retry.OnRetry(func(n uint, err error) {
+			l.Warn("retrying package push",
+				"attempt", n+1,
+				"max_attempts", opts.Retries,
+				"error", err,
+			)
+		}),
+	)
 	if err != nil {
-		return err
+		return ocispec.Descriptor{}, fmt.Errorf("publish failed after retries: %w", err)
```

Comment:
Given that we're logging when we retry, I think it's reasonable to not include "after retries" here. Reads a bit awkward in the default case of one attempt

Reply:
Updated!